_posts/2021-11-23-how-to-dry-your-ref-arch.adoc

```diff
@@ -0,0 +1,606 @@
+It is typical for only the `tag` attribute to be different across environments. In the guide, you would have to repeat
+this configuration across the environments, but with deep merge, you can promote the `repository` and `pull_policy`
+attributes to the common component configuration:
+
+.common component configuration
+[source,hcl]
+----
+inputs = {
+  container_image = {
+    repository  = "gruntwork/aws-sample-app"
+    pull_policy = "IfNotPresent"
+  }
+}
+----
+
+.dev configuration
+[source,hcl]
+----
+include "parent" {
+  path           = "/path/to/parent/configuration"
+  merge_strategy = "deep"
+}
+
+inputs = {
+  container_image = {
+    tag = "v0.0.4"
+  }
+}
+----
+
+.stage configuration
+[source,hcl]
+----
+include "parent" {
+  path           = "/path/to/parent/configuration"
+  merge_strategy = "deep"
+}
+
+inputs = {
+  container_image = {
+    tag = "v0.0.3"
+  }
+}
+----
+
+In this way, you can leverage deep merge to refactor complex nested inputs in your Terragrunt configuration to further
+DRY up the config.
+
+
+[[expose_include]]
+===== Expose `include`
+
+Terragrunt offers the ability to reference values defined in parent configurations through exposed includes. This
+feature is enabled when the `expose = true` attribute is set on the `include` block. Exposed `include` blocks allow the
+child configuration to reference values that are defined in the parent configuration. These values are available with
+the reference `include.LABEL`.
+
+This is most useful to reference `locals` that are defined in the common component configuration to reuse them in the
+child.
+
+For example, if you wanted to test a new version of the module only in dev:
+
+.common component configuration
+[source,hcl]
+----
+locals {
+  source_base_url = "git::git@github.com:gruntwork-io/terraform-aws-service-catalog.git//modules/networking/vpc"
+}
+
+terraform {
+  source = "${local.source_base_url}?ref=v0.65.0"
+}
+----
+
+.dev configuration
+[source,hcl]
+----
+include "parent" {
+  path   = "/path/to/parent/configuration"
+  expose = true
+}
+
+terraform {
+  source = "${include.parent.locals.source_base_url}?ref=v0.66.0"
+}
+----
+
+Note that the availability of values is subject to the
+https://terragrunt.gruntwork.io/docs/etting-started/configuration/#configuration-parsing-order[configuration parsing
+order] of Terragrunt.
+This means that you won't be able to reference later stage values in early stage blocks, like accessing parent `inputs`
+in `locals` or trying to reference a `dependency` in `locals`.
+
+You can work around some of this limitation by packing values in `inputs`. Terragrunt passes inputs to Terraform in a
+way that Terraform ignores input values that do not correspond to an existing variable in the module. For example, if
+you want to expose a reference variable that uses `dependency` blocks, you can create a private input value in the
+parent configuration that references the `dependency`, and access it using exposed `include`:
```

Comment:
I think it is easier to read and follow this if we are a little more verbose, just because holding a lot of state in the mind is inherently hard. So my suggestion is:

```suggestion
way that Terraform ignores input values that do not correspond to an existing variable in the module. For example,
say you want to expose a reference variable that uses the parent configuration's `dependency` blocks.
You can create a private input value in the parent configuration that references the `dependency`, 
like `dependency.vpc.outputs.vpc_id` in the example below. Now this `dependency` output can be used by child
configurations via the parent's `inputs`. Using exposed `include`, the child configuration can reference the parent's
dependency output with `include.parent.inputs._vpc_id`:
```

Reply:
Makes sense! I decided to reword this slightly and take advantage of the examples via interleaving. WDYT of the version her? https://github.com/gruntwork-io/gruntwork-io.github.io/pull/669